src/com/facebook/buck/jvm/kotlin/JarBackedReflectedKotlinc.java

```diff
@@ -0,0 +1,237 @@
+  private final ImmutableSet<Path> compilerClassPath;
+
+  JarBackedReflectedKotlinc(
+      ImmutableSet<Path> compilerClassPath,
+      Iterable<SourcePath> classpath) {
+    this.compilerClassPath = compilerClassPath;
+    this.classpath = ImmutableSortedSet.copyOf(classpath);
+  }
+
+  @Override
+  public KotlincVersion getVersion() {
+    return VERSION;
+  }
+
+  @Override
+  public String getDescription(
+      ImmutableList<String> options,
+      ImmutableSortedSet<Path> javaSourceFilePaths,
+      Path pathToSrcsList) {
+    StringBuilder builder = new StringBuilder("kotlinc ");
+    Joiner.on(" ").appendTo(builder, options);
+    builder.append(" ");
+    builder.append("@").append(pathToSrcsList);
+
+    return builder.toString();
+  }
+
+  @Override
+  public String getShortName() {
+    return "kotlinc";
+  }
+
+  @Override
+  public ImmutableList<String> getCommandPrefix(SourcePathResolver resolver) {
+    throw new UnsupportedOperationException("In memory kotlinc may not be used externally");
+  }
+
+  @Override
+  public int buildWithClasspath(
+      ExecutionContext context,
+      Path outputDirectory,
+      BuildTarget invokingRule,
+      ImmutableList<String> options,
+      ImmutableSortedSet<Path> kotlinSourceFilePaths,
+      Path pathToSrcsList,
+      ImmutableList<String> extraArguments,
+      ImmutableSortedSet<Path> declaredClassPathEntries,
+      Optional<Path> workingDirectory,
+      ProjectFilesystem projectFilesystem)
+      throws InterruptedException {
+
+    String classpath =
+        Joiner.on(File.pathSeparator).join(transform(declaredClassPathEntries, Object::toString));
+
+    ImmutableList.Builder<String> argsBuilder = ImmutableList.<String>builder()
+        .add(INCLUDE_RUNTIME_FLAG)
+        .add(DESTINATION_FLAG)
+        .add(projectFilesystem.resolve(outputDirectory).toString())
+        .addAll(extraArguments)
+        .addAll(transform(
+            kotlinSourceFilePaths,
+            path -> projectFilesystem.resolve(path).toAbsolutePath().toString())
+        );
+
+    if (!classpath.isEmpty()) {
+      argsBuilder
+          .add(CLASSPATH_FLAG)
+          .add(classpath.isEmpty() ? "''" : classpath);
+    }
+
+    ImmutableList<String> args = argsBuilder.build();
+
+    Set<File> compilerIdPaths = compilerClassPath
+        .stream()
+        .map(Path::toFile)
+        .collect(Collectors.toSet());
+
+    try {
+       Object compilerShim = kotlinShims.computeIfAbsent(
+          compilerIdPaths
+              .stream()
+              .map(File::getAbsolutePath)
+              .collect(Collectors.toSet()),
+          k -> loadCompilerShim(context));
+
+      Method compile = compilerShim
+          .getClass()
+          .getMethod("exec", new Class[] {PrintStream.class, String[].class});
+
+      Class<?> exitCodeClass = compilerShim
+          .getClass()
+          .getClassLoader()
+          .loadClass("org.jetbrains.kotlin.cli.common.ExitCode");
```

Comment:
Java reflection already performs caching, a subsequent call to ``loadClass()`` is basically a hash lookup.